06x-new-kafka-roller.md

```diff
@@ -0,0 +1,326 @@
+```
+2. The roller checks if all of the controller nodes are in `NOT_RUNNING` state. Since they are not and `mixed-3` node has `POD_UNRESPONSIVE` reason, it is restarted and waited to have `READY` state. The `mixed-3`'s context becomes:
+```
+         nodeRef: mixed-3/3
+         nodeRoles: controller,broker
+         state: RESTARTED
+         lastTransition: 654987
+         restartReason: POD_UNRESPONSIVE
+         numRestartAttempts: 1
+         numReconfigAttempts: 0
+         numRetries: 0
+```
+3. `mixed-3` state becomes `READY` and since its `numRestartAttempts` is greater than 1, the roller checks the rest of the nodes. 
+4. The roller checks which node is the active controller and finds that `controller-0` is. It then sends a request to the active controller via AdminClient to describe its `controller.quorum.fetch.timeout` config value.
+5. It then considers restarting `controller-1` and checks if the quorum health would be impacted. The operator sends a request to the active controller to describe the quorum replication state. It finds that majority of the follower controllers have caught up with the quorum leader within the `controller.quorum.fetch.timeout.ms`. 
+6. The roller restarts `controller-1` as it has no impact on the quorum health. When it has `READY` state, the roller repeats the quorum check and restarts `controller-2` and then `controller-0`. 
+7. It then considers restarting `mixed-4`, so it performs quorum healthcheck and then availability check. Both check passes therefore `mixed-4` is restarted. The same is repeated for `mixed-5`.
+8. All the controller and mixed nodes have `READY` state and `numRestartAttempts` set to greater than 1. This means, they have been successfuly restarted, therefore the roller considers restarting the broker nodes.
+9. It sends requests to describe all the topic partitions and their `min.insync.replicas` configuration, and the following list of topics is returned:
+```
+topic("topic-A"), Replicas(9, 10, 11), ISR(9, 10), MinISR(2)
+topic("topic-B"), Replicas(6, 7, 8), ISR(6, 7, 8), MinISR(2)
+topic("topic-C"), Replicas(10, 8, 6), ISR(10, 8, 6), MinISR(2)
+topic("topic-D"), Replicas(7, 9, 11), ISR(7, 9, 11), MinISR(2)
+topic("topic-E"), Replicas(6, 10, 11), ISR(6, 10, 11), MinISR(2)
+```
+
+10. The roller batches the nodes that do not have any topic partition in common and the following batches are created: 
+- (11, 8) - `broker-11` and `broker-8` do not share any topic partitions.
+- (7) - `broker-7` and `broker-10` do not share any topic partitions, however topic-A is at min ISR, therefore 10 cannot be restarted and is removed from the batch.
+- (6) - `broker-6` and `broker-9` do not share any topic partitions, however topic-A is at min ISR, therefore 9 cannot be restarted and is removed from the batch.
+
+11. The roller picks the largest batch containing `broker-11` and `broker-8` and restarts them together. It waits for the nodes to have `READY` and then `LEADING_ALL_PREFERRED` state.
```

Comment:
I still think the leading all preferred replicas has the potential of another race condition with CC. Could we made it opt-out by exposing some config that would be enabled by default?

Reply:
We introduced `restartAndPreferredLeaderElectionDelay`, do you think this can be used to prevent the race condition?

Reply:
I think they are for rebalancing the number of leaders hosted on each broker based on the replicas (preferred leader being the first one in the list) , isn't it? Does CC rebalance if broker is shutdown temporarily and comes back but the partitions that it is preferred leader for has not been reassigned to it yet? In another words, does it rebalance based on the ISR, where the leader might not the preferred leader?

Reply:
ah I think I had misunderstood `leader.imbalance.check.interval.seconds` configs and had to remind myself. I guess `restartAndPreferredLeaderElectionDelay` is not so much different than these configurations. However if the roller is the one triggering the leader election, it has its own delay to apply before doing so to avoid triggering it too soon after a restart. If this delay is set to larger than `leader.imbalance.check.interval.seconds`, leader election is already triggered by kafka, therefore roller won't do anything. The operator would only trigger preferred leader election, if the delay is configured shorter than `leader.imbalance.check.interval.seconds` or `auto.leader.rebalance.enable` is set to false. Perhaps the default for the roller's delay should be set to larger than the default of `leader.imbalance.check.interval.seconds`, so we always give kafka the first opportunity to balance the leaders.

Reply:
`leader.imbalance.check.interval.seconds`'s default is 5 minutes, so that would actually prolong the rolling process, especially when restarting node one by one, so setting the default to larger than this wouldn't work actually. I think we can define slightly different purposes for these configurations, roller's purpose is to simply move the partition back to the preferred leader after it was restarted to cause less partition reassignment, whereas kafka is generally balancing out the leaders across all brokers.

Reply:
Yeah, unfortunately these can't be updated dynamically, so we can't adjust them when the rolling is doing its job and then revert.